### PR TITLE
[tf.data] Return unknown cardinality when preserve_cardinality is false

### PR DESCRIPTION
**Pull request by zhuzilin**
Fix bug mentioned in #40107 .

cc @aaudiber 

Thank you for your time on reviewing this pr.

### CODE REVIEW

File: tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op.cc [hunk 1]
```diff
@@ -125,6 +125,8 @@ class MapAndBatchDatasetOp::Dataset : public DatasetBase {
   }
 
   int64 Cardinality() const override {
+    if (!preserve_cardinality_)
```
Comment (aaudiber) [inline tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op.cc:128]: Even though the `if` contains only a single line, please use curly braces to be consistent with the rest of the file.

File: tensorflow/core/kernels/data/experimental/scan_dataset_op.cc [hunk 2]
```diff
@@ -106,7 +106,12 @@ class ScanDatasetOp : public UnaryDatasetOpKernel {
 
     string DebugString() const override { return "ScanDatasetOp::Dataset"; }
 
-    int64 Cardinality() const override { return input_->Cardinality(); }
+    int64 Cardinality() const override {
+      if (preserve_cardinality_)
+        return input_->Cardinality();
+      else
```
Comment (aaudiber) [inline tensorflow/core/kernels/data/experimental/scan_dataset_op.cc:112]: Google C++ style requires using braces when there is an `else` statement: https://google.github.io/styleguide/cppguide.html#Conditionals

Comment (aaudiber) [reply tensorflow/core/kernels/data/experimental/scan_dataset_op.cc:112]: also applies in map_dataset_op and parallel_map_dataset_op

**Review by aaudiber — COMMENTED** [review]
Thanks @zhuzilin! Please also add test cases to https://github.com/tensorflow/tensorflow/blob/27a02f445f768aaa26ade9f30809f82d766a6eed/tensorflow/python/data/experimental/kernel_tests/cardinality_test.py#L90

File: tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op.cc (hunk 1 above)
Comment (zhuzilin) [reply tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op.cc]: fixed.

File: tensorflow/core/kernels/data/experimental/scan_dataset_op.cc (hunk 2 above)
Comment (zhuzilin) [reply tensorflow/core/kernels/data/experimental/scan_dataset_op.cc]: fixed.

**Comment by zhuzilin** [comment]
@aaudiber 
Thank you for your quick review! As for the test, could you tell me is there a way to separate the test for tf1 and tf2? Because the unknown cardinality will only appear in tf1.

**Comment by aaudiber** [comment]
Sorry, I linked to the wrong test file - it's better to update https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/kernel_tests/cardinality_test.py since `cardinality` has graduated out of `/experimental`.

You can write something like the below to test v1-only functionality:

```python

  def _v1_only_test_combinations():
     ...  # similar to https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/kernel_tests/cardinality_test.py#L30

  @combinations.generate(
      combinations.times(combinations.combine(tf_api_version=1, mode=["eager", "graph"]),
              _v1_only_test_combinations())
  def testCardinalityV1Only(self, dataset_fn, expected_result):
    dataset = dataset_fn()
    self.assertEqual(self.evaluate(dataset.cardinality()), expected_result)
```

**Comment by zhuzilin** [comment]
@aaudiber 
Thank you for teaching me how to add test! Could you take another look?

**Review by aaudiber — DISMISSED** [review]
Looks great, thanks!

**Comment by zhuzilin** [comment]
@aaudiber 
The build failure on ubuntu is due to some test for cardinality in c++. I will soon push updates to them.

**Comment by zhuzilin** [comment]
@aaudiber 
The newly updated tests should be able to fix the building failure. Could have another look?

File: tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op_test.cc [hunk 3]
```diff
@@ -316,17 +316,17 @@ DATASET_OUTPUT_SHAPES_TEST_P(MapAndBatchDatasetOpTest, MapAndBatchDatasetParams,
 std::vector<CardinalityTestCase<MapAndBatchDatasetParams>>
 CardinalityTestCases() {
   return {{/*dataset_params=*/MapAndBatchDatasetParams1(),
-           /*expected_cardinality=*/2},
+           /*expected_cardinality=*/-2},
```
Comment (aaudiber) [inline tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op_test.cc:319]: You can use the [kUnknownCardinality](https://github.com/tensorflow/tensorflow/blob/b3925cc9b62775904734eb789da98bd3d0a9dd0b/tensorflow/core/kernels/data/interleave_dataset_op_test.cc#L396) constant for these

**Review by aaudiber — COMMENTED** [review]
Thanks for tracking down the c++ errors.

File: tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op_test.cc (hunk 3 above)
Comment (zhuzilin) [reply tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op_test.cc]: fixed.

**Review by aaudiber — APPROVED** [review]

**Pull request merged**